File: apps/web/src/lib/api/storefront-checkout.ts
import { getApiBaseUrl, ApiError } from "@/lib/api/client";
import { getStorefrontCartToken } from "@/lib/api/storefront-cart";

export type ApiStorefrontCheckoutPayload = {
  email: string;
  first_name: string;
  last_name: string;
  phone: string;
  address: string;
  district?: string;
  postal_code?: string;
  notes?: string;
  payment_method?: string;
  shipping_method?: string;
};

export type ApiStorefrontCheckoutResult = {
  id: string;
  order_number: string;
  email: string;
  grand_total: string;
  payment_method: string;
};

export type ApiStorefrontCheckoutResponse = {
  data: ApiStorefrontCheckoutResult;
};

export async function submitStorefrontCheckout(
  payload: ApiStorefrontCheckoutPayload,
): Promise<ApiStorefrontCheckoutResult> {
  const token = getStorefrontCartToken();
  const res = await fetch(`${getApiBaseUrl()}/api/v1/storefront/checkout`, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
      ...(token ? { "X-Cart-Token": token } : {}),
    },
    body: JSON.stringify(payload),
  });

  if (!res.ok) {
    let detail = res.statusText;
    try {
      const body = (await res.json()) as { detail?: string };
      if (body.detail) detail = body.detail;
    } catch {
      /* ignore */
    }
    throw new ApiError(detail, res.status);
  }

  const data = (await res.json()) as ApiStorefrontCheckoutResponse;
  return data.data;
}

File: apps/api/app/routers/storefront_checkout.py
from __future__ import annotations

from datetime import datetime
from decimal import Decimal

from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session

from app.database import get_db
from app.models.commerce_order import CommerceOrder, dump_json
from app.models.commerce_order_item import CommerceOrderItem
from app.models.storefront_cart import StorefrontCart, StorefrontCartItem
from app.routers.storefront_cart import get_cart_context
from app.schemas.storefront import (
    StorefrontCheckoutCreate,
    StorefrontCheckoutRead,
    StorefrontCheckoutResponse,
)
from app.security import new_session_token

router = APIRouter(prefix="/checkout", tags=["storefront-checkout"])

FREE_SHIPPING_THRESHOLD = Decimal("2000")
SHIPPING_RATES = {
    "standard": Decimal("80"),
    "express": Decimal("150"),
}

PAYMENT_LABELS = {
    "cod": "Cash on Delivery",
    "bkash": "bKash",
    "card": "Credit / Debit Card",
    "bank": "Bank Transfer",
}


def _shipping_amount(method: str, subtotal: Decimal) -> Decimal:
    if subtotal >= FREE_SHIPPING_THRESHOLD:
        return Decimal("0")
    return SHIPPING_RATES.get(method, SHIPPING_RATES["standard"])


def _order_number() -> str:
    suffix = new_session_token()[:8].upper()
    return f"ORD-{suffix}"


@router.post("", response_model=StorefrontCheckoutResponse)
def submit_checkout(
    payload: StorefrontCheckoutCreate,
    cart: StorefrontCart = Depends(get_cart_context),
    db: Session = Depends(get_db),
) -> StorefrontCheckoutResponse:
    if not cart.items:
        raise HTTPException(status_code=400, detail="Cart is empty")

    subtotal = sum((item.unit_price * item.quantity for item in cart.items), Decimal("0"))
    shipping_amount = _shipping_amount(payload.shipping_method, subtotal)
    grand_total = subtotal + shipping_amount

    customer_name = f"{payload.first_name.strip()} {payload.last_name.strip()}".strip()
    payment_key = payload.payment_method.lower()
    payment_method = PAYMENT_LABELS.get(payment_key, payload.payment_method)

    order = CommerceOrder(
        order_number=_order_number(),
        order_date=datetime.utcnow(),
        status="pending",
        payment_status="unpaid",
        shipment_status="unfulfilled",
        source="Web Store",
        customer_name=customer_name,
        customer_phone=payload.phone.strip(),
        customer_email=payload.email.strip().lower(),
        billing_address=payload.address.strip(),
        billing_city=payload.district.strip() or None,
        billing_region=payload.district.strip() or None,
        billing_country="Bangladesh",
        shipping_address=payload.address.strip(),
        shipping_city=payload.district.strip() or None,
        shipping_region=payload.district.strip() or None,
        shipping_country="Bangladesh",
        payment_method=payment_method,
        paid_amount=Decimal("0"),
        due_amount=grand_total,
        shipping_cost=shipping_amount,
        subtotal=subtotal,
        shipping_amount=shipping_amount,
        grand_total=grand_total,
        notes=payload.notes,
        timeline_json=dump_json(
            [
                {
                    "id": "evt_checkout",
                    "type": "created",
                    "label": "Order placed",
                    "at": datetime.utcnow().isoformat(),
                },
            ],
        ),
    )
    db.add(order)
    db.flush()

    for idx, item in enumerate(cart.items):
        line_total = item.unit_price * item.quantity
        db.add(
            CommerceOrderItem(
                order_id=order.id,
                product_id=item.product_id,
                sku=item.slug,
                name=item.name,
                image_url=item.thumbnail,
                quantity=item.quantity,
                unit_price=item.unit_price,
                line_total=line_total,
                sort_order=idx,
            ),
        )

    db.query(StorefrontCartItem).filter(StorefrontCartItem.cart_id == cart.id).delete()
    db.commit()

    return StorefrontCheckoutResponse(
        data=StorefrontCheckoutRead(
            id=order.id,
            order_number=order.order_number,
            email=order.customer_email,
            grand_total=grand_total,
            payment_method=payment_method,
        ),
    )
